processing/src/main/java/org/apache/druid/segment/QueryableIndexCursorSequenceBuilder.java

```diff
@@ -50,51 +56,66 @@
 
 import javax.annotation.Nullable;
 import java.io.IOException;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 
 public class QueryableIndexCursorSequenceBuilder
 {
   private final QueryableIndex index;
   private final Interval interval;
   private final VirtualColumns virtualColumns;
-  @Nullable
-  private final ImmutableBitmap filterBitmap;
+  private final Filter filter;
+  private final QueryMetrics<? extends Query> metrics;
   private final long minDataTimestamp;
   private final long maxDataTimestamp;
   private final boolean descending;
-  @Nullable
-  private final Filter postFilter;
-  @Nullable
-  private final ColumnSelectorColumnIndexSelector bitmapIndexSelector;
 
   public QueryableIndexCursorSequenceBuilder(
       QueryableIndex index,
       Interval interval,
       VirtualColumns virtualColumns,
-      @Nullable ImmutableBitmap filterBitmap,
+      @Nullable Filter filter,
+      @Nullable QueryMetrics<? extends Query> metrics,
       long minDataTimestamp,
       long maxDataTimestamp,
-      boolean descending,
-      @Nullable Filter postFilter,
-      @Nullable ColumnSelectorColumnIndexSelector bitmapIndexSelector
+      boolean descending
   )
   {
     this.index = index;
     this.interval = interval;
     this.virtualColumns = virtualColumns;
-    this.filterBitmap = filterBitmap;
+    this.filter = filter;
+    this.metrics = metrics;
     this.minDataTimestamp = minDataTimestamp;
     this.maxDataTimestamp = maxDataTimestamp;
     this.descending = descending;
-    this.postFilter = postFilter;
-    this.bitmapIndexSelector = bitmapIndexSelector;
   }
 
   public Sequence<Cursor> build(final Granularity gran)
   {
+    final Closer closer = Closer.create();
+
+    // Column caches shared amongst all cursors in this sequence.
+    final ColumnCache columnCache = new ColumnCache(index, closer);
+
     final Offset baseOffset;
 
+    final ColumnSelectorColumnIndexSelector bitmapIndexSelector = new ColumnSelectorColumnIndexSelector(
+        index.getBitmapFactoryForDimensions(),
+        virtualColumns,
+        columnCache
```

Comment:
There's a double-close that can happen as a result of the caching here: `ColumnSelectorColumnIndexSelector.getNumRows` gets the `__time` column and then immediately closes it. So calls to `bitmapIndexSelector.getNumRows()` subsequent to the first one will get an already-closed column. `not` filters call it once, so it'll happen on queries that have multiple `not` filters. And maybe other cases too?

Seems like there's two ways we can fix it:

1. Modify ColumnSelectorColumnIndexSelector to not close the time column, and instead assume it will get closed some other way. Modify places that construct ColumnSelectorColumnIndexSelector to ensure that this happens.

2. Modify places that construct ColumnSelectorColumnIndexSelector to *not* pass in a ColumnCache, but to instead pass in something that doesn't cache columns. Keep the closing behavior of ColumnSelectorColumnIndexSelector the way it is.

Reply:
(2) is not really an option at this point in time.  The whole reason that the column cache is being passed in is because there are a plethora of legitimate bugs where we are not closing columns that need to be closed and passing the columnCache is the way to fix that.  Of course, we could go and change all of the different places where that is broken, but that is a lot more intrusive of a change.

(1) is possible, though, the real fundamental issue is that the `Closer` pattern got added to the `Closeables` that are in the interfaces in Druid.  The addition of the `Closer` meant that we started losing track of where objects are actually being closed as sometimes they are magically closed elsewhere and sometimes they aren't.  We likely need to just have a single pattern (either, it's always via `Closeable`s or it's always via a `Closer`), but the goal of this PR is to get us to where we proactively catch and fix leaks by failing tests, so getting us to a place of consistency in how we deal with closing objects is a bit too big of a thing.

That said, I think I can make it so that the cached Column objects from the columnCache ignore their `close()` method so that the only way to close the underlying object is to do it from the `Closer` which avoids the issue for now.

Reply:
Hrm, actually, no, if I wrap the column objects here, then that will break the contract around the class of object being returned from the `getColumn()`.  We'd need to introduce an `as()` style method to be able to work around that, which is, again, expanding the scope of this PR a bit too much

Reply:
For now, the `__time` column tends to be resilient to double close, so maybe it's okay.

Reply:
nothing to change now, but thinking out loud i wish there was a better way to get the number of rows than checking the time column, its not very accessible and seems a bit overkill for something so useful

Reply:
Based on the conversation here I agree it's OK to leave it as-is, since the `__time` column is currently resilient to double-close, and the changes to avoid it add more scope than we want to bite off right now.

Can you add a comment about this on `ColumnSelectorColumnIndexSelector.getNumRows`, @imply-cheddar, so whoever looks at this code in the future understands what's happening? Something that mentions that the column is being closed here, which will lead to a double-close if it comes from a ColumnCache, but that's fine because the `__time` column is expected to be able to handle a double-close.